test/integrationtests/voight_kampff/test_setup.py

```diff
@@ -0,0 +1,226 @@
+
+    Arguments:
+        msm: msm instance to use for the operations
+        skills: list of skills
+    """
+    for skill_name in skills:
+        skill = msm.find_skill(skill_name)
+        behave_dir = join(skill.path, 'test', 'behave')
+        if exists(behave_dir):
+            copy_feature_files(behave_dir, FEATURE_DIR)
+
+            step_dir = join(behave_dir, 'steps')
+            if exists(step_dir):
+                copy_step_files(step_dir, join(FEATURE_DIR, 'steps'))
+        else:
+            # Generate feature file if no data exists yet
+            print('No feature files exists for {}, '
+                  'generating...'.format(skill_name))
+            # No feature files setup, generate automatically
+            feature = generate_feature(skill_name, skill.path)
+            with open(join(FEATURE_DIR, skill_name + '.feature'), 'w') as f:
+                f.write(feature)
+
+
+def print_install_report(platform, test_skills, extra_skills):
+    """Print in nice format."""
+    print('-------- TEST SETUP --------')
+    yml = yaml.dump({
+        'platform': platform,
+        'test_skills': test_skills,
+        'extra_skills': extra_skills
+        })
+    print(yml)
+    print('----------------------------')
+
+
+def get_arguments(cmdline_args):
+    """Get arguments for test setup.
+
+    Parses the commandline and if specified applies configuration file.
+
+    Arguments:
+        cmdline_args (list): argv like list of arguments
+
+    Returns:
+        Argument parser NameSpace
+    """
+    parser = create_argument_parser()
+    args = parser.parse_args(cmdline_args)
+    return args
+
+
+def create_skills_manager(platform, skills_dir, url, branch):
+    """Create mycroft skills manager for the given url / branch.
+
+    Arguments:
+        platform (str): platform to use
+        skills_dir (str): skill directory to use
+        url (str): skills repo url
+        branch (str): skills repo branch
+
+    Returns:
+        MycroftSkillsManager
+    """
+    repo = SkillRepo(url=url, branch=branch)
+    return MycroftSkillsManager(platform, skills_dir, repo)
+
+
+def main(cmdline_args):
+    """Parse arguments and run test environment setup.
+
+    This installs and/or upgrades any skills needed for the tests and
+    collects the feature and step files for the skills.
+    """
+    args = get_arguments(cmdline_args)
+    if args.config:
+        apply_config(args.config, args)
+
+    msm = create_skills_manager(args.platform, args.skills_dir,
+                                args.repo_url, args.branch)
+
+    random_skills = get_random_skills(msm, args.random_skills)
+    all_skills = args.test_skills + args.extra_skills + random_skills
+
+    install_or_upgrade_skills(msm, all_skills)
+    collect_test_cases(msm, args.test_skills)
+
+    print_install_report(msm.platform, args.test_skills,
+                         args.extra_skills + random_skills)
+
+
+if __name__ == '__main__':
+    main(sys.argv[1:])
```

Comment:
you don't need to pass the command line arguments to the arg parser.  it looks at sys.argv by default.

Reply:
I personally like the traceability. main get's an argument to operate on so all actions can be traced back to it without any questions of where the settings originate.